When a sales order is canceled, when one uses the 'New Copy from Quotati...

guewen opened the pull request:
...on' button,

the binding are moved from the old (canceled) order to the new one, so the
synchronizations now happen on the new order.
- [x] **Warning**: needs this PR to be merged in _addons_ https://github.com/odoo/odoo/pull/1025
- [x]  **Warning**:  needs the same fix in OCB

Original MP on Launchpad: https://code.launchpad.net/~camptocamp/openerp-connector-magento/7.0-copy-quotation-move-binding/+merge/226137

Commit:
When a sales order is canceled, when one uses the 'New Copy from Quotation' button,
the binding are moved from the old (canceled) order to the new one, so the
synchronizations now happen on the new order.

mart-e commented:
The merge in odoo has been accepted in 7.0 at rev odoo/odoo@39f3e40

bwrsandman commented on magentoerpconnect/sale.py:
```diff
@@ -128,6 +128,28 @@ def copy_data(self, cr, uid, id, default=None, context=None):
 
+    def copy_quotation(self, cr, uid, ids, context=None):
+        if isinstance(ids, (tuple, list)):
+            assert len(ids) == 1, "1 ID expected, got %s" % ids
```
Did you mean to put: `% len(ids)` ?

guewen replied:
No, I really meant to put `% ids`, giving more context to the developer needing to debug its code.

gurneyalex replied:
@guewen: careful, there a Python trap here which is going to bite you. If ids is a tuple (and you are testing for this) with more than 1 element (which will sometimes be the case when you reach the assert), then you will get the following exception:

```
TypeError: not all arguments converted during string formatting
```

So this needs to be changed to e.g.:

```
assert len(ids) == 1, "1 ID expected, got the following list %s" % (ids,)
```

guewen replied:
Good catch, thanks

bwrsandman commented:
:+1:

Commit:
Correct Python trap with string formatting

It would try to convert the items of a tuple to arguments of the
string formatting, and would raise the following error:

TypeError: not all arguments converted during string formatting

guewen commented:
Thanks @mart-e for notifying the merge here!

guewen commented:
Rebased from 7.0

guewen commented:
Travis is green on odoo and red on OCB because OCB lacks this fix: odoo/odoo#1025

guewen commented:
Rebased

yvaucher commented:
One test failure (only on OCB):

2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: `Copy a sales order with copy_quotation move bindings
2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: Traceback (most recent call last):
2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module:` File "/home/travis/build/OCA/connector-magento/magentoerpconnect/tests/test_sale_order.py", line 60, in test_sale_order_copy_quotation
2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: `self.assertEqual(mag_line.order_id.id, new_id)
2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module:` AssertionError: 19 != 20L

guewen commented:
> One test failure (only on OCB):
> 
> 2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: Copy a sales order with copy_quotation move bindings 2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: Traceback (most recent call last): 2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: File "/home/travis/build/OCA/connector-magento/magentoerpconnect/tests/test_sale_order.py", line 60, in test_sale_order_copy_quotation
> 2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: self.assertEqual(mag_line.order_id.id, new_id) 2014-08-19 08:42:49,725 23354 TEST openerp_test openerp.modules.module: AssertionError: 19 != 20L

That's what I said on https://github.com/OCA/connector-magento/pull/9#issuecomment-52052200, OCB lacks this fix #125 which is fixed on odoo 7.0.

guewen commented:
Now that OCB is sync'ed again, tests should pass.

nbessi commented:
@guewen is there a risk that duplicated SO is in a state different from original and put synchronisation in a state not supported by connector ?

nbessi commented on magentoerpconnect/sale.py:
```diff
@@ -135,6 +135,29 @@ def copy_data(self, cr, uid, id, default=None, context=None):
+            assert len(ids) == 1, ("1 ID expected, "
+                                   "got the following list %s" % (ids,))
+        if context is None:
+            context = {}
```
We can start using  self.pool['res.users'].context_get(cr, uid) as disscussed on mailing list.
Or maybe it will better to do it on a separated PR

guewen replied:
PR has been done way before it has been discussed but ok I can change that.

nbessi commented on magentoerpconnect/sale.py:
```diff
@@ -202,13 +225,53 @@ class sale_order_line(orm.Model):
 
+    def create(self, cr, uid, vals, context=None):
+        if context is None:
+            context = {}
```
Same comment as before

guewen commented:
@nbessi the state of the sales orders are never pushed to Magento, it was done in the previous connector but was wrong: Magento handles state of the sales orders itself according to the deliveries and invoicing.

nbessi commented:
@guewen perfect :smile:

sebastienbeau commented:
I approve :+1: with the next merge to come #25 the sale order flow will be really complete. Thanks !
